helix-core/src/main/java/org/apache/helix/controller/stages/ManagementMessageGenerationPhase.java

```diff
@@ -0,0 +1,129 @@
+import java.util.Set;
+
+import org.apache.helix.HelixManager;
+import org.apache.helix.api.status.ClusterManagementMode;
+import org.apache.helix.controller.LogUtil;
+import org.apache.helix.controller.dataproviders.ManagementControllerDataProvider;
+import org.apache.helix.controller.pipeline.StageException;
+import org.apache.helix.model.LiveInstance;
+import org.apache.helix.model.LiveInstance.LiveInstanceStatus;
+import org.apache.helix.model.Message;
+import org.apache.helix.model.PauseSignal;
+import org.apache.helix.util.MessageUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Generates participant status change (freeze/unfreeze) and pending state transition cancellation
+ * messages for management mode pipeline.
+ */
+public class ManagementMessageGenerationPhase extends MessageGenerationPhase {
+  private static final Logger LOG = LoggerFactory.getLogger(ManagementMessageGenerationPhase.class);
+
+  @Override
+  public void process(ClusterEvent event) throws Exception {
+    _eventId = event.getEventId();
+    String clusterName = event.getClusterName();
+    HelixManager manager = event.getAttribute(AttributeName.helixmanager.name());
+    ClusterManagementMode managementMode = event.getAttribute(AttributeName.CLUSTER_STATUS.name());
+    ManagementControllerDataProvider cache =
+        event.getAttribute(AttributeName.ControllerDataProvider.name());
+    if (manager == null || managementMode == null || cache == null) {
+      throw new StageException("Missing attributes in event: " + event
+          + ". Requires HelixManager|ClusterStatus|DataCache");
+    }
+
+    PauseSignal pauseSignal = cache.getPauseSignal();
+    if (cache.getClusterConfig().isStateTransitionCancelEnabled()
+        && pauseSignal != null && pauseSignal.getCancelPendingST()) {
+      // Generate ST cancellation messages.
+      LogUtil.logInfo(LOG, _eventId,
+          "Generating ST cancellation messages for cluster " + clusterName);
+      super.process(event);
+    }
+
+    // Is participant status change still in progress? Create messages
+    if (!ClusterManagementMode.Status.COMPLETED.equals(managementMode.getStatus())) {
+      LogUtil.logInfo(LOG, _eventId, "Generating messages as cluster " + clusterName
+          + " is still in progress to change participant status");
+      List<Message> messages =
+          generateStatusChangeMessages(managementMode.getMode(), cache.getEnabledLiveInstances(),
+              cache.getLiveInstances(), cache.getAllInstancesMessages(), manager.getInstanceName(),
+              manager.getSessionId());
+      MessageOutput messageOutput =
+          event.getAttributeWithDefault(AttributeName.MESSAGES_ALL.name(), new MessageOutput());
+      messageOutput.addStatusChangeMessages(messages);
+      event.addAttribute(AttributeName.MESSAGES_ALL.name(), messageOutput);
+    }
+  }
+
+  private List<Message> generateStatusChangeMessages(ClusterManagementMode.Type managementMode,
```

Comment:
When you generate it, you should add them to cache. Not let dispatch phase to cache it.